src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,191 @@
+        for (T element : list) {
+            addLast(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            return getByIndex(index).item;
+        }
+        throw new IndexOutOfBoundsException("Index is out of bounds");
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T previousValue;
+        if (index < size && index >= 0) {
+            previousValue = getByIndex(index).item;
+            getByIndex(index).item = value;
+            return previousValue;
+        }
+        throw new IndexOutOfBoundsException("Index is out of bounds");
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            T removedElement;
+            if (index == 0) {
+                removedElement = head.item;
+                removeFirst();
+                return removedElement;
+            } else if (index == size - 1) {
+                removedElement = tail.item;
+                removeLast();
+                return removedElement;
+            }
+            return removeElement(getByIndex(index));
+        }
+        throw new IndexOutOfBoundsException("Index is out of bounds");
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> current = head;
+        while (current.item == null ? !(current.item == object)
+                : !current.item.equals(object)) {
+            if (current.next == null) {
+                return false;
+            }
+            current = current.next;
+        }
+        if (current == head) {
+            removeFirst();
+            return true;
+        }
+        removeElement(current);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean addLast(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            size++;
+            return true;
+        }
+        Node<T> newNode = new Node<>(tail, value, null);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+        return true;
+    }
+
+    private void addFirst(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            size++;
+            return;
+        }
+        Node<T> newNode = new Node<>(null, value, head);
+        head.previous = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void insert(Node<T> current, T value) {
+        Node<T> newElement = new Node<>(current.previous, value, current);
+        current.previous.next = newElement;
+        current.previous = newElement;
+        size++;
+    }
```

Comment:
```suggestion
    private void insert(Node<T> current, T value) {
        final Node<T> previousNode = current.previous;
        Node<T> newElement = new Node<>(previousNode, value, current);
        current.previous = newElement;
        if (previousNode == null) {
            head = newElement;
        } else {
            previousNode.next = newElement;
        }
        size++;
    }
```

And we can remove addFirst() method.

Reply:
Це не правильно , адже при додававні першого елемента у список нам також потрібно встановлювати хвіст, крім того цей код може викинути NullPointerException. Я вважаю краще розділити обовязки.